### PR TITLE
feat(wmg): Message to users that data underlying GE is changing #4163

### PR DESCRIPTION
**Pull request by ashin-czi**
## Reason for Change

- #4163 

## Changes

- add
    - Message banner for download modal
    - Message in copy link notification
    - Banner in all exports
- remove
- modify
    - Reducing "magic numbers" for calculating heatmap width/height for PNG and SVG
    - Notification from Toast to SDS Notification
    - Fixed Notification on close animation
    - Legend is now centered for png export under the data message

## Testing steps

1. Select tissue(s)
2. Select gene(s)
3. Click download button
4. See new banner in modal
5. Download SVG, PNG, and/or CSV
6. Make sure banner message appears in some form in all exports
7. Ensure that no data is missing or cut off from exports

## Notes for Reviewer

https://data-message-frontend.rdev.single-cell.czi.technology/gene-expression

### CODE REVIEW

File: frontend/src/views/WheresMyGene/components/Main/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/Main/style.ts:21]: BetaWrapper no longer needed as we're wrapping in Beta component instead

File: frontend/src/views/WheresMyGene/components/Main/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/Main/index.tsx:403]: No more BetaWrapper, moved to Beta component

File: frontend/tests/features/wheresMyGene.test.ts
Comment (ashin-czi) [inline frontend/tests/features/wheresMyGene.test.ts:1]: Change to only extract string to constants

File: frontend/src/views/WheresMyGene/components/Main/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/Main/index.tsx:356]: Used for PNG and SVG exports to render message banner to render in output

this will be hidden behind the loading screen and disappear on download complete

File: frontend/src/views/WheresMyGene/components/HeatMap/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/HeatMap/style.ts:89]: Removing this because it introduced a bug where PNGs were getting cut off

File: frontend/src/views/WheresMyGene/components/HeatMap/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/HeatMap/style.ts:23]: Fixes a bug where PNG was getting cut off

File: frontend/src/views/WheresMyGene/components/HeatMap/components/YAxisChart/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/HeatMap/components/YAxisChart/index.tsx:198]: Using this for SVG tissue label rendering

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/ShareButton/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/ShareButton/style.ts:31]: Need to wrap Notification in order for animations to work properly, cannot be set on the Notification component itself

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/ShareButton/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/ShareButton/index.tsx:101]: Migration from Toast to SDS Notification

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/ShareButton/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/ShareButton/index.tsx:57]: No more Toast

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/svgUtils.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/svgUtils.ts:417]: Moving renderDots function from SaveExport/index.tsx to SaveExport/svgUtils.ts

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/svgUtils.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/svgUtils.ts:315]: This fixes a bug where the tissue label was not aligned properly in the SVG

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/svgUtils.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/svgUtils.ts:1]: A lot of these changes are to clean some stuff up and not use "magic numbers" to align things, and instead using existing constant variables to calculate positioning.

Also making use of the `applyAttributes` function instead of `setAttribute` as it's much cleaner

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/style.ts:1]: Can be ignored, just styling changes

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:466]: creating object parameters

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:452]: no more hardcoded number

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:345]: moved to SaveExport/svgUtils.ts

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:369]: using `applyAttributes` from svgUtils instead of using `setAttribute` multiple times

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:335]: Getting the svg banner that is rendered in the UI behind the loading screen on download. This svg element will be inserted into the final svg as-is

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:318]: renaming from processSvg to generateSvg for consistency with generateCsv cause it's easier for me to remember

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:291]: The banner will appear above the download button in the modal. Setting width to 100% here to fill the width of the modal

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:255]: using styled div now for additional modal styling

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:219]: The observer listens to changes in the dom, and once detected it will be triggered and then the download still start after `MUTATION_OBSERVER_TIMEOUT` seconds. We only do this for PNG and CSV because `echartsRendererMode` only needs to be set for SVG, and changing the renderer mode also triggers the download.

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:177]: we're listening for changes in attributes now so that this line:
`heatmapNode.classList.add("is-downloading");`

triggers the download without having the set the echarts renderer for all filetypes.

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:171]: we need to change the observer target node since we now render the message banner while downloading, in order for the banner to render in the exports. The banner is a child of the container with id="view", so adding and removing the banner will trigger the observer twice, thus causing the file to be downloaded twice.

We only want to listen to changes to the heatmap anyways, so we have the observer listen to the container with id `heatmap-container-id`

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/ExportBanner/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/ExportBanner/index.tsx:1]: This is a new component which holds the svg element of the data banner. Have to do this because our config somewhere doesn't allow us to import an svg string directly. Instead it gets converted to a Next object.

This also gives us the ability to set the width/height of the svg in places we need, like the download modal

File: frontend/src/views/WheresMyGene/components/Beta/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/Beta/index.tsx:11]: Wrapping the notification in this component now instead of in main/index.tsx

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (tihuan) [reply frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:219]: I think @seve was saying PNG quality is better when echarts is in SVG mode vs. PNG mode? If that's true, then I think only CSV download needs this hack!

**Review by tihuan — COMMENTED** [review]
Probably only have time to review PR on Monday, but just wanted to thank you for all the comments! Super high quality PR that's a really easy to follow and review 🔥

File: frontend/src/views/WheresMyGene/components/Beta/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/Beta/style.ts:10]: targetting specific id because using the generic `.elevated` class was messing with the copy link notification

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/ExportBanner/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/ExportBanner/index.tsx:1]: This is a new component which holds the svg element of the data banner. Have to do this because our config somewhere doesn't allow us to import an svg string directly. Instead it gets converted to a Next object.

This method also gives us the ability to set the width/height of the svg in places we need through props, like the download modal

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/csvUtils.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/csvUtils.ts:57]: New change to add data change message after the timestamp row

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:263]: Still waiting on captions from product

**Comment by ashin-czi** [comment]
Moving this back to draft because we're still waiting for the caption text in the download modal

image

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/style.ts:1]: Mostly just styling changes for the download modal

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:298]: Decided by UX team that text should only be centered for exports

File: frontend/src/views/WheresMyGene/components/HeatMap/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/HeatMap/index.tsx:176]: Do not need to check if fmg open anymore for calculating viewbox width, fixed here: https://github.com/chanzuckerberg/single-cell-data-portal/pull/4463/files#r1153767695

File: frontend/src/views/WheresMyGene/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/style.ts:30]: Removing some styles from UI only for export to center items

File: frontend/src/views/WheresMyGene/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/style.ts:1]: Only changes are extracting variables to constants

File: frontend/tests/features/wheresMyGene.test.ts
Comment (ashin-czi) [inline frontend/tests/features/wheresMyGene.test.ts:1]: https://github.com/chanzuckerberg/single-cell-data-portal/pull/4463/files#r1153767695

**Review by seve — COMMENTED** [review]
Intitial stale review, submitting now and taking another look later today

File: frontend/src/views/WheresMyGene/components/Beta/index.tsx
Comment (ashin-czi) [reply frontend/src/views/WheresMyGene/components/Beta/index.tsx:13]: Could you give an example of how to use the constant? Not sure what you're referring to exactly!

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/ExportBanner/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/ExportBanner/index.tsx:35]: Text is rendered as an svg path element, so we have separate path data depending on the context

File: frontend/src/components/common/Filter/common/style.ts
Comment (ashin-czi) [inline frontend/src/components/common/Filter/common/style.ts:64]: Creating a base NotifcationWrapper for beta and share url notification wrapper

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/style.ts:1]: Only styling changes for download modal

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/ShareButton/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/ShareButton/style.ts:1]: Only styling changes for share url Notification

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/ShareButton/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/ShareButton/style.ts:38]: Increasing the size of the url icon since SDS only has small sizes for the icon

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (tihuan) [reply frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:300]: add to code comment please!

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:351]: great comments!

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx
Comment (tihuan) [reply frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/index.tsx:452]: big win!!

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/svgUtils.ts
Comment (tihuan) [reply frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveExport/svgUtils.ts:1]: Thanks for refactoring this!

File: frontend/src/views/WheresMyGene/components/HeatMap/style.ts
Comment (tihuan) [reply frontend/src/views/WheresMyGene/components/HeatMap/style.ts:89]: I don't have time to test this - but I'm assuming the heatmap scrollbar won't be covered by the right sidebar after removing this?

**Review by tihuan — APPROVED** [review]
LGTM! Just some comments I think are worth adding to code 🚀 !

Might need another reviewer's help to test the app, I don't have time right now sorry!

**Pull request merged**
